### PR TITLE
Removed `refreshCompSpeeds` + removed automatic comp claims + added `_setCompSpeed`

### PR DESCRIPTION
**Pull request by Pet3ris**
**WORK IN PROGRESS**

**Authors**:

- @pluriholonomic, @jmomx, @reichiang, @pet3ris

**Key features**:

- Remove `refreshCompSpeeds` and replace with manual `_setCompSpeed` for each market
- Remove automatic COMP claims

**Other changes**:

- Remove `_addCompMarket` and `_dropCompMarket` inlining them in `_setCompSpeed` instead
- Remove `transferComp` and use `grantCompInternal` instead
- Deprecate `compClaimThreshold`
- Deprecate `compRate` and remove `_setCompRate` and `NewCompRate` event
- Deprecate `isComped`
- Refactor distribution functions
- Fixed gas leak introduced in proposal 30 and discovered by jvaqa https://www.comp.xyz/t/compound-contributor-grants/756/7
- Fixed gas reporting test
- Added `yarn report` command to print out a gas report (going forward, developers can run `yarn profile && yarn report` to get a report

**Testnet testing plan**

Set comp speed:
S1 - COMP speed can be set per market
S2 - COMP speed can be changed per market
S3 - COMP speed can be set to 0 and market removed
S4 - COMP rate division within a market is unchanged
S5 - Attempting to call `_setCompRate` fails

Disabled COMP claims:
A1 - COMP is not claimed on protocol interactions
A2 - COMP can still be claimed separately

General Comptroller consistency:
C1 - Check COMP speed consistency pre/post patch for markets
C2 - Check mint works correctly
C3 - Check redeem works correctly
C4 - Check borrow works correctly
C5 - Check repay works correctly
C6 - Check liquidate works correctly

**Auditing note**

Most of this patch is a subset of the audited vesting patch, however, the amendments to `_addCompMarket` and `_dropCompMarket`, inclusions into `_setCompSpeed`, refactoring of distribution and removal of `isComped` were done after the audit and should be carefully examined by the community.

**Remaining steps**
- [x] Finish test amendments
- [x] Audit
- [x] Compound team review
- [x] Deployment scenarios
- [x] Deploy to testnet
- [x] Complete testnet testing
- [x] Deploy to mainnet
- [x] Submit proposal

### CODE REVIEW

File: spec/sim/0006-setspeed-manual-claims/hypothetical_upgrade.scen [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+MineBlock
+From CompHolder (Governor GovernorAlpha Proposal LastProposal Vote For)
+AdvanceBlocks 20000
+Governor GovernorAlpha Proposal LastProposal Queue
+IncreaseTime 604910
+Governor GovernorAlpha Proposal LastProposal Execute
+
+-- Check new speed
+Assert Equal (Comptroller CompSpeed cUSDC) 2.7e16
```
Comment (jflatow) [inline spec/sim/0006-setspeed-manual-claims/hypothetical_upgrade.scen:101]: Are there any other checks that Comp is still being distributed appropriately (and ideally other market fns) that you can do?

File: contracts/Comptroller.sol [hunk 2]
```diff
@@ -1277,12 +1252,12 @@ contract Comptroller is ComptrollerV5Storage, ComptrollerInterface, ComptrollerE
      * @dev Note: If there is not enough COMP, we do not perform the transfer all.
      * @param user The address of the user to transfer COMP to
      * @param amount The amount of COMP to (possibly) transfer
-     * @return The amount of COMP which was NOT transferred to the user
+     * @return The amount of COMk which was NOT transferred to the user
```
Comment (coburncoburn) [inline contracts/Comptroller.sol:1255]: let's stick with calling it COMP here, not possible to change the token symbol at this point in time.

**Comment by Pet3ris** [comment]
Gas statistics:

```
Gas report
----------

first mint - 129978 gas
second mint - 90506 gas
second mint, no interest accrued - 66475 gas
redeem - 96782 gas
unitroller-g6 second mint with comp accrued - 145034 gas
unitroller-g6 claim comp - 142390 gas
unitroller second mint with comp accrued - 124317 gas
unitroller claim comp - 150385 gas
```

For more context, unitroller-g6 is the previous patch implementation while unitroller is the current patch. The other scenarios are benchmarked only for current match but didn't change between patches. For precise definition of the scenarios, please refer to `gasProfiler.js`

File: spec/sim/0006-setspeed-manual-claims/hypothetical_upgrade.scen (hunk 1 above)
Comment (Pet3ris) [reply spec/sim/0006-setspeed-manual-claims/hypothetical_upgrade.scen:100]: There are now 3 checks:

- USDCWhale accrual changes when the proposal changes USDC flow
- DAIWhale pre & post patch consistency check with supply holdings
- BigBorrower pre & post patch consistency check with both supply and borrow holdings in different tokens

I've added a new file `hypothetical_upgrade_borrower.scen` with the new borrowing check.

**Comment by Pet3ris** [comment]
Added `post_deploy` and `post_propose` scripts in a new commit now that proposal is submitted.

For reference, the commit at the time of proposal was 2b75b97eeb3211c02d4948860faf26eac7e69d6e.

**Comment by Pet3ris** [comment]
@jflatow @coburncoburn ready for final review post merge https://compound.finance/governance/proposals/33.

**Review by jflatow — APPROVED** [review]

**Pull request merged**